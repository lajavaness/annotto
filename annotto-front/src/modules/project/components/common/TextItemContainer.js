import { isEmpty } from 'lodash'
import PropTypes from 'prop-types'
import React, { useMemo } from 'react'

import { sanitizer } from 'shared/utils/htmlUtils'

import * as Styled from './__styles__/TextItemContainer.styles'

const TextItemContainer = ({ content, highlights }) => {
	const resolveContentWithHighlight = useMemo(() => {
		let result = [{ value: content, isHighlight: false }]
		let lastEnd = 0

		highlights.forEach(({ start, end, text }) => {
			const lastValue = result[result.length - 1].value
			result[result.length - 1] = { value: lastValue.slice(0, start - lastEnd), isHighlight: false }
			result.push({ value: text, isHighlight: true })
			result.push({ value: lastValue.slice(end - lastEnd), isHighlight: false })
			lastEnd = end
		})

		return result
	}, [highlights, content])

	const sanitizedHtml = useMemo(() => sanitizer(content), [content])

	return (
		<Styled.Root>
			{!isEmpty(highlights) ? (
				<Styled.Content>
					{resolveContentWithHighlight.map(({ value, isHighlight }, index) =>
						isHighlight ? (
							<Styled.Mark data-testid={'__mark__'} key={index}>
								{value}
							</Styled.Mark>
						) : (
							<span key={index}>{value}</span>
						)
					)}
				</Styled.Content>
			) : (
				<Styled.Content dangerouslySetInnerHTML={{ __html: sanitizedHtml }} />
			)}
		</Styled.Root>
	)
}

export default TextItemContainer

TextItemContainer.propTypes = {
	/** Defines the content of the item. */
	content: PropTypes.string,
	/** Defines the highlights of the item. */
	highlights: PropTypes.arrayOf(
		PropTypes.shape({
			/** Defines the character where the highlight begins. */
			start: PropTypes.number,
			/** Defines the character where the highlight ends. */
			end: PropTypes.number,
			/** Defines the text of the highlight. */
			text: PropTypes.string,
			/** Defines the score of the highlight. The value is between 0 and 1. */
			score: PropTypes.number,
		})
	),
}

TextItemContainer.defaultProps = {
	content: null,
	highlights: [],
}
