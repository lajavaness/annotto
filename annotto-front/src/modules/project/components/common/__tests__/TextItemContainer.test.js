import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import TextItemContainer from '../TextItemContainer'

const defaultProps = {}
const getInstance = (props = {}) => (
	<ThemeProvider theme={theme}>
		<TextItemContainer {...defaultProps} {...props} />
	</ThemeProvider>
)

describe('TextItemContainer', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})

	it('renders content', () => {
		const content = 'Foo'
		const { getByText } = render(getInstance({ content }))
		expect(getByText(content)).toBeVisible()
	})

	it('renders highlights', () => {
		const text = 'Foo'
		const content = `${text} Bar`
		const highlights = [{ start: 0, end: 3, text, score: 0.2 }]
		const { getByTestId } = render(getInstance({ content, highlights }))
		expect(getByTestId('__mark__')).toBeVisible()
		expect(getByTestId('__mark__')).toHaveTextContent(text)
	})
})
