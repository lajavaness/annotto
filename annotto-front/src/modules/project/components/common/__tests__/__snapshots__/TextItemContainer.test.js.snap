// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextItemContainer matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsBqHK bmaRKC"
  >
    <p
      class="sc-aKXVD"
    />
  </div>
</DocumentFragment>
`;
